www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,290 @@
+- Transaction prepared successfully but did not commit. In this case the coordinator checks the results of the transaction vote in its in-memory state, adding a "yes" vote from the recovered database. Based on the overall vote, the coordinator then decides if the transaction should commit or roll-back and distributes commands accordingly. 
+- Transaction does not appear in either of the database state tables. This means that this database already rolled back the transaction, and it never successfully prepared. This means that the database voted "no" in the commit vote, and the transaction coordinator has to distribute `ROLLBACK PREPARED` to the other databases that participated in the transaction.
+
+There is one special case for database failure, which is the case where the database where the transaction originated has failed. If you recall, this database is responsible for notifying the transaction coordinator that about the `XACT_EVENT_PRE_COMMIT` and `XACT_EVENT_COMMIT` events for the transaction. If the originating database fails, the coordinator may be stuck with an "orphan" transactions that never get prepared or committed. For this case, we rely on the transaction timeouts and a scheduled thread on the coordinator to roll back the orphaned transaction.
+
+### Recovering from coordinator failure
+
+If the coordinator itself fails, the first order of business is to reconstruct the in-memory state, so it will be aware of all the on-going transactions and capable of completing them. It first uses the latest state from the `dist_tx_participants`, and then start checking state of individual databases.
+For each transaction, each database can either have it in committed state, in prepared state, or not have it at all. This leads to a few possible scenarios:
+
+- Transaction is in a committed state in all databases. In this case, we are done and can update the in-memory state and metadata accordingly.
+- Transaction is in prepared state in some databases and committed in others. This indicates that the vote was successful, but the commit wasn't fully distributed. The coordinator will distribute `COMMIT PREPARED` to the databases that are still in prepared state and complete the transaction.
+- Transaction is in prepared state in some databases and does not exist in others. This indicates a failed vote. The databases where the transaction does not exist have rolled it back already. The coordinator will distribute `ROLLBACK PREPARED` to the remaining databases. 
+- Transaction is in committed state in some databases and does not exist in others. This should never ever happen. If some databases committed, it indicates a unanimous vote to commit, but the databases without the transaction have voted "no" and rolled back. This shouldn't happen, and hasn't happened in months of both intensive testing and production use.
```

Comment:
We can mention the most tricky case to handle - when ALL participants are in PREPARED state. In this case, we have a special procedure to decide what to do with the transaction. In this case we use 'distributed_transactions' table on the originating PDB. If transaction is in the table - decision commit, no transaction - decision roll back.

Reply:
If the transaction is in `distributed_transactions` on the originating PDB, it means that it was already committed, which means that it is in "prepared" state on most PDBs and in "committed" state in the originating, which is already covered. 

It sounds like the tricky part is that the originating DB doesn't have a real `prepared` state since we do normal commit there (not two phase). So when we observe the originating PDB is can only be in "committed" or "rollback" state. 

For the sake of simplifying the explanation, I'll leave this out. Both here and in the earlier comment on  "doomed to succeed".

Reply:
the transaction can also still be in progress, so we don't know yet if it's after the precommit phase and whether it can still fail or not.

and one of the problem is that the transaction id exposed is 32bits, which can wrap round.  So it can be tricky to detect whether a transaction X still in progress is actually the same transaction X that was distributed (which we store as a full transaction id, which cannot wraparound).

Reply:
The transaction can still be in progress on the originating PDB while the remote PDBs are in "prepared" state. TC just distributed the "prepare transaction" to remote PDBs, got OK from all of them, updated the status to PREPARED in the participants table for every remote PDB. At this moment the transaction on the originating PDB is still in-progress. Imagine TC crashes, then it restarts, get the PREPARED state for all participants, but the transaction on the originating PDB still could be in-progress....